text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+extensions and layers.
+As an example, if an assertion depends on the `sed` utility. OSX ships by default with BSD sed,
+while Linux ships with GNU sed, which have subtly different behaviours and option switches.
+In the proposed solution, the assertion will either use the `sed` utility in the Lambda base
+image or bring its own via Lambda Layers.
+If this is not accounted for, this alternative will cause the assertion to behave inconsistently.
+
+[AWS CloudFormation custom resources]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html
+[cloud assembly]: https://docs.aws.amazon.com/cdk/api/latest/docs/cloud-assembly-schema-readme.html#cloud-assembly
+[AWS Lambda base images]: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-images.html
+[Modifying runtime environment]: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-modify.html
+
+## Appendix B - Test Execution
+
+### Design
+
+The AWS CDK CLI (aka toolkit) is responsible for deploying and destroying AWS CDK apps.
+We will extract these two parts from the CLI into a separate jsii module.
+
+The `run()` method on the class `apptest.Test` will depend on this module to deploy and
+destroy stacks as part of test runs.
+
+Once the CDK app and assertions have finished deploying (deploying assertions implies
+execution of the assertions - see Appendix A), the `run()` method will query the status
+of the stack in CloudFormation.
+
+If the stack is not in one of the expected `COMPLETE` states, an error (or `Exception`
+in some languages) will be thrown containing the relevant failure messages.
+The failure messages will be sourced from messages from the AWS CloudFormation stack
+and the log stream produced by the test run.
+
+### Alternative - custom test executor
+
+An alternate approach is to ship our own test framework & execution system
+as part of the AWS CDK CLI.
+
+An example design for such an approach will consist -
+
+A project set up for apptest will have the `apptest` section in its `cdk.json`.
+In the case of a javascript CDK app:
+
+```json
+{
+  "apptest": {
+    "exec": "node {}",
+    "testRegex": "**/*.apptest.js"
+  }
+}
+```
+
+The apptest suite for a project is executed by running `cdk apptest`.
+
+When invoked, it will discover all apptest test cases using the `testRegex` provided.
+For each match, run synthesis using the value provided in the `exec` option, followed
+by deployment and finally destruction of the stack and assertions.
+
+At the end of the run, the CLI will download the detailed test results from AWS Logs to the
+location `apptest.out/results-xxx`.
+
+The major downside to this solution is that it is re-inventing another test execution mechanism.
+
+Users will now need to learn how to organize apptest tests, learn how to execute it and read
+results. We are not leveraging existing "well known" test frameworks.
+
+Although this looks basic today, as the scope of apptest expands and more features are added,
```

Comment:
I understand the desire to reuse execution&reporting from existing frameworks but I am not sure this makes total sense because users won't be able to use any assertion functions and the programming experience will likely be confusion.

As an exercise, can you provide a more complete example/prototype of a jest-based test suite that uses this? I am curious how this will look?